From: Egbert Eich <eich@suse.de>
Date: Thu, 30 Jun 2011 10:18:22 +0200
Subject: VGA arb: Only enable legacy resources if device really decodes them.
Patch-mainline: to be mainlined

If a device doesn't decode legacy resources there is no need to
acquire them for the device.

Signed-off-by: Egbert Eich <eich@suse.de>
Signed-off-by: Takashi Iwai <tiwai@suse.de>
---
 drivers/gpu/vga/vgaarb.c |   14 +++++++++-----
 1 file changed, 9 insertions(+), 5 deletions(-)

--- linux-3.0-SLE11-SP2-3.0.orig/drivers/gpu/vga/vgaarb.c
+++ linux-3.0-SLE11-SP2-3.0/drivers/gpu/vga/vgaarb.c
@@ -175,6 +175,9 @@ static struct vga_device *__vga_tryget(s
 	    (vgadev->decodes & VGA_RSRC_LEGACY_MEM))
 		rsrc |= VGA_RSRC_LEGACY_MEM;
 
+	/* only turn on legacy resources if device decodes them */
+	rsrc &= ~VGA_RSRC_LEGACY_MASK | vgadev->decodes;
+
 	pr_debug("%s: %d\n", __func__, rsrc);
 	pr_debug("%s: owns: %d\n", __func__, vgadev->owns);
 
@@ -578,12 +581,13 @@ static inline void vga_update_device_dec
 		vgadev->owns &= ~old_decodes;
 		list_for_each_entry(new_vgadev, &vga_list, list) {
 			if ((new_vgadev != vgadev) &&
-			    (new_vgadev->decodes & VGA_RSRC_LEGACY_MASK)) {
+			    (new_vgadev->decodes & old_decodes & VGA_RSRC_LEGACY_MASK)) {
 				pr_info("vgaarb: transferring owner from PCI:%s to PCI:%s\n", pci_name(vgadev->pdev), pci_name(new_vgadev->pdev));
-				conflict = __vga_tryget(new_vgadev, VGA_RSRC_LEGACY_MASK);
-				if (!conflict)
-					__vga_put(new_vgadev, VGA_RSRC_LEGACY_MASK);
-				break;
+				conflict = __vga_tryget(new_vgadev, old_decodes & VGA_RSRC_LEGACY_MASK);
+				if (!conflict) {
+					__vga_put(new_vgadev, old_decodes & VGA_RSRC_LEGACY_MASK);
+					break;
+				}
 			}
 		}
 	}
